fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -407,20 +407,358 @@ class ReadIteratorPool {
 	uint64_t iteratorsReuseCount;
 };
 
-class PerfContextMetrics {
+ACTOR Future<Void> flowLockLogger(const FlowLock* readLock, const FlowLock* fetchLock) {
+	loop {
+		wait(delay(SERVER_KNOBS->ROCKSDB_METRICS_DELAY));
+		TraceEvent e("RocksDBFlowLock");
+		e.detail("ReadAvailable", readLock->available());
+		e.detail("ReadActivePermits", readLock->activePermits());
+		e.detail("ReadWaiters", readLock->waiters());
+		e.detail("FetchAvailable", fetchLock->available());
+		e.detail("FetchActivePermits", fetchLock->activePermits());
+		e.detail("FetchWaiters", fetchLock->waiters());
+	}
+}
+
+struct DataShard {
+	DataShard(std::string name) : name(name), writeBatch(std::make_unique<rocksdb::WriteBatch>()) {}
+
+	~DataShard() {
+		if (db == nullptr)
+			return;
+		// Close DB
+		readIterPool.reset();
+		auto s = db->Close();
+		if (!s.ok()) {
+			logShardEvent(name, ShardOp::CLOSE, SevError, s.ToString());
+			logRocksDBError(s, "CloseShard");
+			return;
+		}
+		logShardEvent(name, ShardOp::CLOSE);
+
+		if (!deletePending || segments.size() > 0)
+			return;
+
+		std::vector<rocksdb::ColumnFamilyDescriptor> defaultCF = { rocksdb::ColumnFamilyDescriptor{ "default",
```

Comment:
Do we need to destroyDB each time a DataShard is destructed? That will delete the RocksDB from the disk?